runtime/sak.c

```diff
@@ -0,0 +1,148 @@
+     represented as a C string literal.
+
+     On Unix, `sak encode-C-literal /usr/local/lib` returns `"/usr/local/lib"`
+
+     On Windows, `sak encode-C-literal "C:\OCaml🐫\lib"` returns
+     `L"C:\\OCaml\xd83d\xdc2b\\lib"`
+   - add-stdlib-prefix. Used in stdlib/StdlibModules to convert the list of
+     basenames given in STDLIB_MODULE_BASENAMES to the actual file basenames
+     in STDLIB_MODULES.
+
+     For example, `sak add-stdlib-prefix stdlib camlinternalAtomic Sys` returns
+     ` stdlib camlinternalAtomic stdlib__Sys`
+ */
+
+void usage(void)
+{
+  printf(
+    "OCaml Build System Swiss Army Knife\n"
+    "Usage: sak command\n"
+    "Commands:\n"
+    " * encode-C-literal path - encodes path as a C string literal\n"
+    " * add-stdlib-prefix name1 ... - prefix standard library module names\n"
+  );
+}
+
+/* Converts the supplied path (UTF-8 on Unix and UCS-2ish on Windows) to a valid
+   C string literal. On Windows, this is always a wchar_t* (L"..."). */
+void encode_C_literal(char_os *path)
+{
+  char_os c;
+
+#ifdef _WIN32
+  putchar('L');
+#endif
+  putchar('"');
+
+  while ((c = *path++) != 0) {
+    /* Escape \, " and \n */
+    if (c == '\\') {
+      printf("\\\\");
+    } else if (c == '"') {
+      printf("\\\"");
+    } else if (c == '\n') {
+      printf("\\n");
+#ifndef _WIN32
+    /* On Unix, nothing else needs escaping */
+    } else {
+      putchar(c);
+#else
+    /* On Windows, allow 7-bit printable characters to be displayed literally
+       and escape everything else (using the older \x notation for increased
+       compatibility, rather than the newer \U. */
+    } else if (c < 0x80 && iswprint(c)) {
+      putwchar(c);
+    } else {
+      printf("\\x%04x", c);
+#endif
+    }
+  }
+
+  putchar('"');
+}
+
+/* Print the given array of module names to stdout. "stdlib" and names beginning
+   "camlinternal" are printed unaltered. All other names are prefixed "stdlib__"
+   with the original name capitalised (i.e. "foo" prints "stdlib__Foo"). */
+void add_stdlib_prefix(int count, char_os **names)
+{
+  int i;
+  char_os *name;
+
+  for (i = 0; i < count; i++) {
+    name = *names++;
+
+    /* "stdlib" and camlinternal* do not get changed. All other names get
+       capitalised and prefixed "stdlib__". */
+    if (strcmp_os(T("stdlib"), name) == 0
+     || strncmp_os(T("camlinternal"), name, 12) == 0) {
+      printf_os(T(" %s"), name);
+    } else {
+      /* name is a null-terminated string, so an empty string simply has the
+         null-terminator "capitalised". */
+      *name = toupper_os(*name);
+      printf_os(T(" stdlib__%s"), name);
+    }
+  }
+}
+
+#ifdef _WIN32
+int wmain(int argc, wchar_t **argv)
+#else
+int main(int argc, char **argv)
+#endif
+{
+  if (argc == 3 && !strcmp_os(argv[1], T("encode-C-literal"))) {
+    encode_C_literal(argv[2]);
+  } else if (argc > 1 && !strcmp_os(argv[1], T("add-stdlib-prefix"))) {
+    add_stdlib_prefix(argc - 2, &argv[2]);
```

Comment:
Isn't `&argv[2]` just `argv + 2`?

Reply:
Yes, they're identical (including the code generated). To me the `&argv[2]` more clearly "looks" like an array with the first two elements removed than the pointer arithmetic version, but I have no strong opinion!